### DISCUSSION TITLE
The router is unable to add Starlette's HTTPEndpoint

### DISCUSSION THREAD

**Discussion by didip** [fastapi/fastapi#8142] [Questions] [answered]
**Describe the bug**
I cannot find a way to add Starlette's HTTPEndpoint onto app.router.

**To Reproduce**

1. Create app: `app = fastapi.FastAPI()`

2. Create an endpoint:

```
class HealthHandler(starlette.endpoints.HTTPEndpoint):
    async def get(self, request):
        return starlette.responses.JSONResponse({'status': 'OK'})
```

3. I am stuck now, I cannot see a way to add HTTPEndpoint into `app.router`.

**Expected behavior**
Able to add HTTPEndpoint class.

**Environment:**

- OS: macOS

- FastAPI Version: `fastapi==0.29.1`
 
- Python version: 3.7.3

**Comment by stefanondisponibile** [comment, marked as answer]
That's not really [how FastAPI's meant to look like](https://fastapi.tiangolo.com/tutorial/first-steps/).
However, if you wanted to define an endpoint that way, remember [you are in a Starlette environment](https://fastapi.tiangolo.com/features/#starlette-features):

> `FastAPI` is actually a sub-class of `Starlette`.

Therefore, what you could do is:

```python
# main.py
from fastapi import FastAPI
from starlette.endpoints import HTTPEndpoint
from starlette.responses import JSONResponse

app = FastAPI()

@app.route("/")
class HealthHandler(HTTPEndpoint):
    async def get(self, request):
        return JSONResponse({'status': 'OK'})
```

```bash
uvicorn main:app
# INFO: Started server process [19630]
# INFO: Waiting for application startup.
# INFO: Uvicorn running on http://127.0.0.1:8000 (Press CTRL+C to quit)
# INFO: ('127.0.0.1', 50608) - "GET / HTTP/1.1" 200 ==> {"status":"OK"}
```

**Comment by didip** [comment]
Follow up question if you don't mind,

what if the Endpoint class is defined in a different module, how can I add it without using the decorator syntax?

**Comment by stefanondisponibile** [comment]
One solution could be this:

```python
#health_handler.py

from main import app
from starlette.endpoints import HTTPEndpoint
from starlette.responses import JSONResponse

@app.route("/")
class HealthHandler(HTTPEndpoint):
    async def get(self, request):
        return JSONResponse({'status': 'OK'})
```

```python
# main.py
from fastapi import FastAPI

app = FastAPI()
import health_handler
```

**Comment by HarrySky** [comment]
@stefanondisponibile @didip  `Starlette` (and `FastAPI` as subclass) has `add_route` method so there is no need to create cyclical dependencies:

```
#health_handler.py

from starlette.endpoints import HTTPEndpoint
from starlette.responses import JSONResponse

class HealthHandler(HTTPEndpoint):
    async def get(self, request):
        return JSONResponse({'status': 'OK'})
```

```
# main.py
from fastapi import FastAPI
from health_handler import HealthHandler

app = FastAPI()
app.add_route(path="/", route=HealthHandler)
```

**Comment by didip** [comment]
@HarrySky follow up question, if I do `app.add_route`, I don't get the auto generated Swagger, yes?

Is there a `app.add_api_route` that also generate Swagger?

**Comment by tiangolo** [comment]
Yep. @stefanondisponibile is correct. And then @HarrySky's insight is correct too :smile: 

Thanks for the help @stefanondisponibile and @HarrySky !

**Comment by tiangolo** [comment]
@didip The Swagger autogeneration depends on OpenAPI.

OpenAPI depends on being able to know the parameters of your request. And that currently depends on using the normal decorator style, not the `HTTPEndpoint` style.

**Reply by Judekeyser** [reply]
Personally I'd love to see it becoming something, because the class approach of Starlette is sometimes much more idiomatic in some Restful perceptions (especially when dealing with HATEOAS style). Having only the path-based points makes it look often too much like RPC and that is sometimes difficult to work around the design decision.

**Comment by didip** [comment]
Gotcha, thank you for all the explanations, guys!

**Comment by tiangolo** [comment]
Thanks for reporting back and closing the issue.
